app/utils/context.tsx

```diff
@@ -1,90 +1,222 @@
 "use client";
-import { WalletState } from "@/hooks/useWalletConnection";
+import { ConnectedWalletsState } from "@/hooks/useWalletConnection";
 import React, {
   createContext,
   useContext,
   useState,
   useEffect,
   ReactNode,
 } from "react";
+import { ProviderKey, SUPPORTED_WALLETS, WalletKey } from "./wallet/constants";
+import { GetConnectorsReturnType } from "wagmi/actions";
+import { connectEVMWallet, connectUTXOWallet } from "./wallet/walletConnect";
 
 interface AppStateContextType {
   isWalletModalOpen: boolean;
   toggleWalletModal: () => void;
-  wallet: WalletState | null;
-  setWalletState: (wallet: WalletState) => void;
+  walletsState: ConnectedWalletsState | null;
+  setWalletsState: React.Dispatch<React.SetStateAction<ConnectedWalletsState>>;
+  getProviderTypeFromChain: (chain: string) => ProviderKey;
 }
 
 const AppStateContext = createContext<AppStateContextType | undefined>(
-  undefined,
+  undefined
 );
 
 export const AppStateProvider = ({ children }: { children: ReactNode }) => {
   const [isWalletModalOpen, setIsWalletModalOpen] = useState(false);
-  const [wallet, setWalletState] = useState<WalletState | null>(null);
+  const [walletsState, setWalletsState] = useState<ConnectedWalletsState>({});
 
   const toggleWalletModal = () => {
     setIsWalletModalOpen((prevState) => !prevState);
   };
 
-  useEffect(() => {
-    const provider = wallet?.provider;
-
-    if (!provider) {
-      console.warn(
-        "Provider does not support event listeners. Consider adding polling for updates.",
-      );
-      return;
+  const getProviderTypeFromChain = (chain: string): ProviderKey => {
+    switch (chain) {
+      case "AVAX":
+      case "BSC":
+      case "ETH": {
+        return ProviderKey.EVM;
+      }
+      case "BTC": {
+        return ProviderKey.BITCOIN;
+      }
+      case "DOGE": {
+        return ProviderKey.DOGECOIN;
+      }
     }
+    return ProviderKey.EVM;
+  };
 
-    if (typeof provider.on === "function") {
-      const handleAccountsChanged = (accounts: string[]) => {
-        if (!wallet) return;
-        setWalletState({ ...wallet, address: accounts[0] || "" });
-      };
+  const checkAvailableWallets = (window: any) => {
```

Comment:
Is there any reason not to declare the connect functions on SUPPORTED_WALLETS? A lot of this stuff doesn't rely on changing app state.

Perhaps you would want to do this because the window.provider might be injected at app runtime? That wouldn't make sense since checkAvailableWallets executes on first render.